File: Lab01/Lexer.cs
/*
 Example language fragment:
   var x = 2;
   print x + 5;
   print "hello world";

 ops: + - * /
 strings: "..."
 keywords: var print if else while
*/

namespace Lab01;

public sealed class Lexer
{
    private readonly string _input;
    private readonly int _length;
    private int _position;

    public Lexer(string input)
    {
        _input = input;
        _length = input.Length;
        _position = 0;
    }

    public List<Token> Tokenize()
    {
        var result = new List<Token>();

        while (_position < _length)
        {
            var current = Peek();

            if (char.IsWhiteSpace(current))
            {
                Next();
                continue;
            }

            if (char.IsDigit(current))
            {
                TokenizeNumber(result);
                continue;
            }

            if (char.IsLetter(current))
            {
                TokenizeWord(result);
                continue;
            }

            if (current == '"')
            {
                TokenizeString(result);
                continue;
            }

            TokenizeOperator(result);
        }

        return result;
    }

    private void TokenizeNumber(List<Token> result)
    {
        var start = _position;

        while (char.IsDigit(Peek()))
            Next();

        var numberStr = _input.Substring(start, _position - start);
        result.Add(new Token(TokenType.NUMBER, numberStr, start));
    }

    private void TokenizeWord(List<Token> result)
    {
        var start = _position;

        while (char.IsLetterOrDigit(Peek()))
            Next();

        var word = _input.Substring(start, _position - start);

        switch (word)
        {
            case "var":
                AddToken(result, TokenType.VAR, word, start);
                break;
            case "print":
                AddToken(result, TokenType.PRINT, word, start);
                break;
            case "if":
                AddToken(result, TokenType.IF, word, start);
                break;
            case "else":
                AddToken(result, TokenType.ELSE, word, start);
                break;
            case "while":
                AddToken(result, TokenType.WHILE, word, start);
                break;
            default:
                AddToken(result, TokenType.ID, word, start);
                break;
        }
    }

    /// <summary>Double-quoted string: <c>"hello world"</c> → value is the inner text (no quotes).</summary>
    private void TokenizeString(List<Token> result)
    {
        var start = _position;
        Next(); // opening "

        var contentStart = _position;
        while (_position < _length && Peek() != '"')
            Next();

        if (_position >= _length)
            throw new InvalidOperationException($"Unterminated string starting at position {start}");

        var inner = _input.Substring(contentStart, _position - contentStart);
        Next(); // closing "

        result.Add(new Token(TokenType.STRING, inner, start));
    }

    private void TokenizeOperator(List<Token> result)
    {
        var current = Peek();
        var start = _position;

        switch (current)
        {
            case '+':
                Next();
                AddToken(result, TokenType.PLUS, "+", start);
                break;
            case '-':
                Next();
                AddToken(result, TokenType.MINUS, "-", start);
                break;
            case '*':
                Next();
                AddToken(result, TokenType.STAR, "*", start);
                break;
            case '/':
                Next();
                AddToken(result, TokenType.SLASH, "/", start);
                break;
            case '=':
                Next();
                AddToken(result, TokenType.EQ, "=", start);
                break;
            case ';':
                Next();
                AddToken(result, TokenType.SEMICOLON, ";", start);
                break;

            default:
                throw new InvalidOperationException(
                    $"Unexpected character '{current}' at position {_position}");
        }
    }

    /// <summary>Current character without consuming it.</summary>
    private char Peek()
    {
        if (_position >= _length)
            return '\0';

        return _input[_position];
    }

    /// <summary>Consume and return current character.</summary>
    private char Next()
    {
        if (_position >= _length)
            return '\0';

        return _input[_position++];
    }

    private static void AddToken(List<Token> result, TokenType type, string value, int start) =>
        result.Add(new Token(type, value, start));
}

File: src/mini_compiler/lexer.py
from __future__ import annotations

from .token import Token
from .token_type import TokenType


class Lexer:
    def __init__(self, input_text: str) -> None:
        self._input = input_text
        self._length = len(input_text)
        self._position = 0

    def tokenize(self) -> list[Token]:
        result: list[Token] = []

        while self._position < self._length:
            current = self._peek()

            if current.isspace():
                self._next()
                continue

            if current.isdigit():
                self._tokenize_number(result)
                continue

            if current.isalpha():
                self._tokenize_word(result)
                continue

            if current == '"':
                self._tokenize_string(result)
                continue

            self._tokenize_operator(result)

        result.append(Token(TokenType.EOF, "", self._length))
        return result

    def _tokenize_number(self, result: list[Token]) -> None:
        start = self._position
        while self._peek().isdigit():
            self._next()
        number_str = self._input[start : self._position]
        result.append(Token(TokenType.NUMBER, number_str, start))

    def _tokenize_word(self, result: list[Token]) -> None:
        start = self._position
        while self._peek().isalnum():
            self._next()
        word = self._input[start : self._position]

        if word == "var":
            self._add_token(result, TokenType.VAR, word, start)
        elif word == "print":
            self._add_token(result, TokenType.PRINT, word, start)
        elif word == "if":
            self._add_token(result, TokenType.IF, word, start)
        elif word == "else":
            self._add_token(result, TokenType.ELSE, word, start)
        elif word == "while":
            self._add_token(result, TokenType.WHILE, word, start)
        else:
            self._add_token(result, TokenType.ID, word, start)

    def _tokenize_string(self, result: list[Token]) -> None:
        start = self._position
        self._next()  # opening "

        content_start = self._position
        while self._position < self._length and self._peek() != '"':
            self._next()

        if self._position >= self._length:
            raise RuntimeError(f"Unterminated string starting at position {start}")

        inner = self._input[content_start : self._position]
        self._next()  # closing "

        result.append(Token(TokenType.STRING, inner, start))

    def _tokenize_operator(self, result: list[Token]) -> None:
        current = self._peek()
        start = self._position

        if current == "(":
            self._next()
            self._add_token(result, TokenType.LPAREN, "(", start)
        elif current == ")":
            self._next()
            self._add_token(result, TokenType.RPAREN, ")", start)
        elif current == "{":
            self._next()
            self._add_token(result, TokenType.LBRACE, "{", start)
        elif current == "}":
            self._next()
            self._add_token(result, TokenType.RBRACE, "}", start)
        elif current == "+":
            self._next()
            self._add_token(result, TokenType.PLUS, "+", start)
        elif current == "-":
            self._next()
            self._add_token(result, TokenType.MINUS, "-", start)
        elif current == "*":
            self._next()
            self._add_token(result, TokenType.STAR, "*", start)
        elif current == "/":
            self._next()
            self._add_token(result, TokenType.SLASH, "/", start)
        elif current == ";":
            self._next()
            self._add_token(result, TokenType.SEMICOLON, ";", start)
        elif current == "=":
            if self._peek(1) == "=":
                self._next()
                self._next()
                self._add_token(result, TokenType.EQEQ, "==", start)
            else:
                self._next()
                self._add_token(result, TokenType.EQ, "=", start)
        elif current == "!":
            if self._peek(1) == "=":
                self._next()
                self._next()
                self._add_token(result, TokenType.NEQ, "!=", start)
            else:
                self._next()
                self._add_token(result, TokenType.EXCL, "!", start)
        elif current == "<":
            if self._peek(1) == "=":
                self._next()
                self._next()
                self._add_token(result, TokenType.LTEQ, "<=", start)
            else:
                self._next()
                self._add_token(result, TokenType.LT, "<", start)
        elif current == ">":
            if self._peek(1) == "=":
                self._next()
                self._next()
                self._add_token(result, TokenType.GTEQ, ">=", start)
            else:
                self._next()
                self._add_token(result, TokenType.GT, ">", start)
        elif current == "&":
            if self._peek(1) == "&":
                self._next()
                self._next()
                self._add_token(result, TokenType.AND, "&&", start)
            else:
                raise RuntimeError(f"Unexpected character '{current}' at position {self._position}")
        elif current == "|":
            if self._peek(1) == "|":
                self._next()
                self._next()
                self._add_token(result, TokenType.OR, "||", start)
            else:
                raise RuntimeError(f"Unexpected character '{current}' at position {self._position}")
        else:
            raise RuntimeError(f"Unexpected character '{current}' at position {self._position}")

    def _peek(self, offset: int = 0) -> str:
        i = self._position + offset
        if i >= self._length:
            return "\0"
        return self._input[i]

    def _next(self) -> str:
        if self._position >= self._length:
            return "\0"
        ch = self._input[self._position]
        self._position += 1
        return ch

    @staticmethod
    def _add_token(result: list[Token], typ: TokenType, value: str, start: int) -> None:
        result.append(Token(typ, value, start))
